src/braket/default_simulator/cirq/cirq_program_context.py

```diff
@@ -0,0 +1,122 @@
+from typing import List, Tuple
+
+import cirq
+import numpy as np
+from braket.ir.jaqcd.program_v1 import Results
+from cirq import (
+    Circuit,
+    bit_flip,
+    PhaseFlipChannel,
+    DepolarizingChannel,
+    AmplitudeDampingChannel,
+    GeneralizedAmplitudeDampingChannel,
+    PhaseDampingChannel,
+)
+
+from braket.default_simulator.openqasm.parser.braket_pragmas import (
+    AbstractBraketPragmaNodeVisitor,
+    parse_braket_pragma,
+)
+from braket.default_simulator.openqasm.parser.generated.BraketPragmasParser import (
+    BraketPragmasParser,
+)
+from braket.default_simulator.openqasm.program_context import AbstractProgramContext
+
+# cirq.XX, cirq.YY, and cirq.ZZ gates are not the same as Braket gates
+CIRQ_GATES = {
+    "i": cirq.I,
+    "h": cirq.H,
+    "x": cirq.X,
+    "y": cirq.Y,
+    "z": cirq.Z,
+    "cnot": cirq.CNOT,
+    "cz": cirq.CZ,
+    "s": cirq.S,
+    "t": cirq.T,
+    "cphaseshift": cirq.cphase,
+    "rx": cirq.rx,
+    "ry": cirq.ry,
+    "rz": cirq.rz,
+    "swap": cirq.SWAP,
+    "iswap": cirq.ISWAP,
+    "ccnot": cirq.CCNOT,
+    "cswap": cirq.CSWAP,
+    "ccz": cirq.CCZ,
+    "ccx": cirq.CCX,
+    "measure": cirq.MeasurementGate,
+}
+
+
+class CirqProgramContext(AbstractProgramContext):
+    def __init__(self):
+        super().__init__(Circuit())
+
+    def is_builtin_gate(self, name: str) -> bool:
+        user_defined_gate = self.is_user_defined_gate(name)
+        return name in CIRQ_GATES and not user_defined_gate
+
+    def add_phase_instruction(self, target: Tuple[int], phase_value: int):
+        raise NotImplementedError
+
+    def _gate_accepts_parameters(self, gate_class):
+        return hasattr(gate_class, "parameters") and len(gate_class.parameters) > 0
+
+    def add_gate_instruction(
+        self, gate_name: str, target: Tuple[int], params, ctrl_modifiers: List[int], power: int
+    ):
+        qubits = [cirq.LineQubit(int(qubit)) for qubit in target]
+        target_qubits = qubits[len(ctrl_modifiers) :]
+        control_qubits = qubits[: len(ctrl_modifiers)]
+
+        if params:
+            gate = CIRQ_GATES[gate_name](*params).on(*target_qubits)
+        else:
+            gate = CIRQ_GATES[gate_name].on(*target_qubits)
+        ctrl_modifiers = [bit ^ 1 for bit in ctrl_modifiers]
+        gate = gate.controlled_by(*control_qubits, control_values=ctrl_modifiers)
+        gate = gate**power
+
+        self.circuit.append(gate)
+
+    def add_custom_unitary(
+        self,
+        unitary: np.ndarray,
+        target: Tuple[int],
+    ) -> None:
+        """Add a custom Unitary instruction to the circuit"""
+        raise NotImplementedError
+
+    def add_noise_instruction(self, noise):
+        """Add a noise instruction the circuit"""
+        self.circuit.append(noise)
```

Comment:
This looks the same as the base class? Do we need to overwrite here?

Reply:
In the abstract base class this method is defined as abstract method.

Since adding the gate to a circuit changes based on the circuit we are using.

In BraketProgramContext we do it as circuit.add_instruction

In CirqProgramContext we do it as circuit.append